src/swarm-tui/Swarm/TUI/Controller/EventHandlers/Main.hs

```diff
@@ -0,0 +1,155 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+-- |
+-- SPDX-License-Identifier: BSD-3-Clause
+--
+-- Here is the main player configurable key event handler while playing the game.
+module Swarm.TUI.Controller.EventHandlers.Main (
+  mainEventHandlers,
+) where
+
+import Brick
+import Brick.Keybindings
+import Control.Lens as Lens
+import Control.Monad (unless, void, when)
+import Control.Monad.IO.Class (liftIO)
+import Swarm.Game.Scenario.Topography.Structure.Recognition (automatons)
+import Swarm.Game.Scenario.Topography.Structure.Recognition.Type (originalStructureDefinitions)
+import Swarm.Game.State
+import Swarm.Game.State.Substate
+import Swarm.Game.Step (finishGameTick)
+import Swarm.TUI.Controller.EventHandlers.Frame (runGameTickUI)
+import Swarm.TUI.Controller.UpdateUI (updateUI)
+import Swarm.TUI.Controller.Util
+import Swarm.TUI.Editor.Model (isWorldEditorEnabled, worldOverdraw)
+import Swarm.TUI.Model
+import Swarm.TUI.Model.Event (MainEvent (..), SwarmEvent (..))
+import Swarm.TUI.Model.Goal
+import Swarm.TUI.Model.UI
+import System.Clock (Clock (..), TimeSpec (..), getTime)
+
+-- | Main keybindings event handler while running the game itself.
+--
+-- See 'Swarm.TUI.Controller.handleMainEvent'.
+mainEventHandlers :: [KeyEventHandler SwarmEvent (EventM Name AppState)]
+mainEventHandlers = allHandlers Main $ \case
```

Comment:
So just to make sure I understand the process, let's say we wanted to add a new action which can be bound to a key by the user, but doesn't have a keybinding by default.  It looks like all we would have to do is add a constructor to one of the enumeration types in `Swarm.TUI.Model.Event`, and then update a few functions (such as here) wherever the compiler yells at you about missing patterns.  Is that it, or is there anything else?

Reply:
Yes, if it is part of existing handlers, you would add it to `Swarm.TUI.Model.Event` and leave the default list of bindings empty.

There are two less common situations that require more work:
* If the event has a condition _and_ is also bound in sub-handlers, then you add it to the old-style handler.
* If you need a new dispatcher (for example, to add the main scenarios menu), then it has to be added to more places as described in `Swarm.TUI.Controller.EventHandlers`.

Reply:
Would be good to make sure this info ends up in a comment somewhere obvious, for anyone who wants to add some handlers in the future.

Reply:
I am not sure there is such an obvious place while `Controller` is still used for some handlers.

Maybe we could use the note reference in all relevant handlers? 🤔

Reply:
> Maybe we could use the note reference in all relevant handlers? 🤔

Yes, that's a good idea.